### PR TITLE
[FLINK-32608][table-planner] Support scan reusing with projection pushdown

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

*Currently, Flink doesn't support scan source reuse when one source table is referenced by different projection pushdown. For this problem, we can merge all projections in one source, then the source can be reused. So this PR proposes to support scan reusing with multiple different projection pushdown*

## Brief change log

  - *Support scan reusing with multiple different projection pushdown*
  - *Add Plan test to verify the changes*

## Verifying this change

This change added tests and can be verified as follows:
  - *Added unit tests  in ScanReuseTest*
  - *Added integration test in ScanReuseITCase*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Review by lincoln-lil — CHANGES_REQUESTED** [review]
@lsyldliu thanks for contributing this! I've left some comments, PTAL.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/FlinkCalcMergeRule.scala [hunk 1]
```diff
@@ -65,6 +66,23 @@ class FlinkCalcMergeRule[C <: Calc](calcClass: Class[C])
     val topCalc: Calc = call.rel(0)
     val bottomCalc: Calc = call.rel(1)
 
+    val newCalc = merge(topCalc, bottomCalc)
```
[the remark this answers is not in the document]
Comment (lincoln-lil) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/FlinkCalcMergeRule.scala:69]: It's my oversight.. this `merge` method should be moved into `FlinkRelUtil`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/ScanReuser.java [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            if (watermarkSpec.isPresent()) {
+                specs.add(watermarkSpec.get());
+                newSourceType = watermarkSpec.get().getProducedType().get();
+            }
+
+            // 2.3 Create a new ScanTableSource. ScanTableSource can not be pushed down twice.
+            Map<String, String> mergedOptions =
+                    new HashMap<>(pickTable.contextResolvedTable().getTable().getOptions());
+            mergedOptions.putAll(HintUtil.convertRelHintToMap(pickScan.getHints()));
+            // TODO how to handle different hints for reusable scans
```
[the remark this answers is not in the document]
Comment (lincoln-lil) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/ScanReuser.java:207]: Also add a case: one source was queried multiple times via different table hints

[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/ScanReuser.java]: Yes, you are right, I have added a test to verify it.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/ScanReuser.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            // 2.2 Watermark spec
+            Optional<WatermarkPushDownSpec> watermarkSpec =
+                    getAdjustedWatermarkSpec(pickTable, newSourceType);
+            if (watermarkSpec.isPresent()) {
+                specs.add(watermarkSpec.get());
+                newSourceType = watermarkSpec.get().getProducedType().get();
+            }
+
+            // 2.3 Create a new ScanTableSource. ScanTableSource can not be pushed down twice.
+            Map<String, String> mergedOptions =
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/ScanReuser.java]: After some rethinking, I also think we don't need this logic. But we also don't need the validation logic, it looks strange here. The digest has guaranteed different hints doesn't be reused.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/spec/DynamicTableSourceSpec.java [hunk 4]
```diff
@@ -71,12 +72,25 @@ public DynamicTableSourceSpec(
...
     private DynamicTableSource getTableSource(FlinkContext context, FlinkTypeFactory typeFactory) {
         if (tableSource == null) {
-            final DynamicTableSourceFactory factory =
+            DynamicTableSourceFactory factory =
                     context.getModuleManager()
                             .getFactory(Module::getTableSourceFactory)
                             .orElse(null);
 
             ResolvedCatalogTable resolvedCatalogTable = contextResolvedTable.getResolvedTable();
+            if (factory == null) {
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/spec/DynamicTableSourceSpec.java:81]: Yes, this change proposed is to fix the hive `testQueries` and `testAdditionalQueries` test. I have split this change into a separated commit.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/FlinkCalcMergeRule.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/FlinkCalcMergeRule.scala]: We need it, the merge logic can be reused by `ReplaceScanWithCalcShuttle`. I have extracted it to util method to `FlinkRelUtil`

**Comment by lsyldliu** [comment]
@lincoln-lil Thanks for your review, I've updated it.

**Review by lincoln-lil — APPROVED** [review]
LGTM

**Comment by lincoln-lil** [comment]
The failed test is tracked by FLINK-32632.

**Comment by twalthr** [comment]
@lsyldliu @lincoln-lil for next time: could we split such big PRs into smaller ones? It is very difficult for reviewers to review 4000 lines with the appropriate quality. This PR got my attention because I saw unnecessary calls to `CatalogManager` in `DynamicTableSourceSpec`. I'm pretty sure there are more code simplifications that fell under the table.

**Pull request merged**